smolyay/surrogate.py

```diff
@@ -1,546 +1,717 @@
+        # reset constant
+        self._integration_constant = 0
+        if not self._terms_constructed_cache:
+            self._create_terms()
+            self._terms_constructed_cache = True
+
+        # get points
+        if isinstance(
+            X,
+            (
+                smolyay.samples.UnidimensionalPointSet,
+                smolyay.samples.MultidimensionalPointSet,
+            ),
+        ):
+            X = X.points
+
+        # validate data inputs
+        X = numpy.array(X,ndmin=2,copy=None)
+        self._in_domain(X)
+
+        y = numpy.asarray(y)
+        if y.shape != (X.shape[0],) and y.shape != (X.shape[0], 1):
+            raise IndexError("Must be 1D array with shape (n_samples,).")
+
+        # create lookup table and solve for all the basis functions
+        lookup_table = [
+            self.basis_sets[dim](X[:, dim], self.domain[dim])
+            for dim in range(self.num_dimensions)
+        ]
+
+        # create basis matrix
+        if any(
+            any(bf._is_complex for bf in basis_set) for basis_set in self.basis_sets
+        ):
+            basis_matrix = numpy.zeros(
+                (len(X), len(self._index_combinations)), dtype=complex
+            )
+        else:
+            basis_matrix = numpy.zeros((len(X), len(self._index_combinations)))
+
+        # use lookup table to solve for each term
+        for term, ic in enumerate(self._index_combinations):
+            basis_matrix[:, term] = numpy.prod(
+                [lookup_table[dim][ic[dim], :] for dim in range(len(ic))], axis=0
+            )
+
+        # solve for coefficients
+        if self.regularization is None or numpy.any(numpy.iscomplex(basis_matrix)):
+            if basis_matrix.shape[0] == basis_matrix.shape[1]:
+                try:
+                    self._coefficients = numpy.linalg.solve(basis_matrix, y)
+                except numpy.linalg.LinAlgError:
+                    self._coefficients = numpy.linalg.lstsq(
+                        basis_matrix, y, rcond=None
+                    )[0]
             else:
-                value = basis(points)
-            basis_matrix[:,j]= value
-
-        if linear_solver == 'lu':
-            self._coefficients = numpy.linalg.solve(
-                basis_matrix,
-                self._data)
-        elif linear_solver == 'inv':
-            self._coefficients = numpy.dot(numpy.linalg.inv(
-                basis_matrix),
-                self._data)
-        elif linear_solver == 'lstsq':
-            self._coefficients = numpy.linalg.lstsq(basis_matrix,
-                                                    self._data,
-                                                    rcond=None)[0]
+                self._coefficients = numpy.linalg.lstsq(basis_matrix, y, rcond=None)[0]
+        elif isinstance(self.regularization, L2Regularization):
+            regressor = sklearn.linear_model.Ridge(fit_intercept = False,**self.regularization.__dict__)
```

Comment:
Don't unpack the whole object, be explicit about what you want to forward:
```suggestion
            regressor = sklearn.linear_model.Ridge(alpha=self.regularization.alpha, fit_intercept=False)
```